File: backend/app/schemas/tracker.py
"""求职进度 Schema。

识别导入走"**先预览、再确认**"两步：解析接口只算不写，把"这条会新增 / 会更新 /
没变化"连同原因一起返回；用户确认后再提交真正的写入。这与岗位多份导入是同一条
原则——识别结果在用户核对之前不落库。

写入路径都不接受 `company_key` / `title_key`：合并键由服务层从公司名与岗位名算出，
不让调用方（包括模型）有机会构造一个"看起来一样但合并不到一起"的键。
"""
from __future__ import annotations

import re
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from .extraction import (
    MAX_EXTRACTION_DOCUMENT_COUNT,
    MAX_EXTRACTION_IMAGE_COUNT,
    ExtractionDocumentInput,
    ExtractionImageInput,
)

MAX_TRACK_COMPANY_CHARS = 128
MAX_TRACK_TITLE_CHARS = 128
MAX_TRACK_STAGE_NOTE_CHARS = 64
MAX_TRACK_NOTE_CHARS = 4_000
MAX_TRACK_EVIDENCE_CHARS = 500
MAX_TRACK_NEXT_ACTION_CHARS = 255
MAX_TRACK_DATE_CHARS = 16
# 一次最多处理多少条识别结果：一封邮件通常只对应一条，一次粘贴几十条已经很多了。
MAX_TRACK_RECORDS = 30
MAX_TRACK_TEXT_CHARS = 50_000

_DATE_RE = re.compile(r"\A\d{4}-\d{2}-\d{2}\Z")


def _is_date_or_empty(value: str) -> bool:
    """``YYYY-MM-DD`` 或空串，且必须是真实存在的日历日（挡掉 2026-02-31）。"""
    if not value:
        return True
    if not _DATE_RE.match(value):
        return False
    try:
        date.fromisoformat(value)
    except ValueError:
        return False
    return True


class TrackRecordIn(BaseModel):
    """一条进度记录的内容（识别结果或手工填写），不含 id 与合并键。"""

    model_config = ConfigDict(extra="forbid")

    company: str = Field(default="", max_length=MAX_TRACK_COMPANY_CHARS)
    title: str = Field(default="", max_length=MAX_TRACK_TITLE_CHARS)
    status: str = Field(default="applied", max_length=24)
    stage_note: str = Field(default="", max_length=MAX_TRACK_STAGE_NOTE_CHARS)
    applied_at: str = Field(default="", max_length=MAX_TRACK_DATE_CHARS)
    status_date: str = Field(default="", max_length=MAX_TRACK_DATE_CHARS)
    next_action: str = Field(default="", max_length=MAX_TRACK_NEXT_ACTION_CHARS)
    next_action_date: str = Field(default="", max_length=MAX_TRACK_DATE_CHARS)
    note: str = Field(default="", max_length=MAX_TRACK_NOTE_CHARS)
    # 这一条是从哪句话看出来的。只存简短片段，不存整封邮件。
    evidence: str = Field(default="", max_length=MAX_TRACK_EVIDENCE_CHARS)

    @field_validator("status")
    @classmethod
    def status_must_be_known(cls, value: str) -> str:
        from ..models.tracker import STATUSES

        cleaned = (value or "").strip()
        if cleaned not in STATUSES:
            raise ValueError(f"未知的进度状态「{cleaned}」")
        return cleaned

    @field_validator("applied_at", "status_date", "next_action_date")
    @classmethod
    def dates_must_be_real(cls, value: str) -> str:
        cleaned = (value or "").strip()
        if not _is_date_or_empty(cleaned):
            raise ValueError("日期要写成 2026-09-18 这样的真实日期，或留空")
        return cleaned

    @model_validator(mode="after")
    def must_identify_a_position(self) -> "TrackRecordIn":
        """没有公司或没有岗位，就无从判断它该并到哪一条上。"""
        self.company = self.company.strip()
        self.title = self.title.strip()
        self.stage_note = self.stage_note.strip()
        self.next_action = self.next_action.strip()
        self.note = self.note.strip()
        self.evidence = self.evidence.strip()
        if not self.company or not self.title:
            raise ValueError("请填写公司和岗位名称——进度要按这两项合并")
        return self


class TrackCreate(TrackRecordIn):
    """手工新建一条进度。"""

    job_id: int | None = Field(default=None, ge=1)
    resume_id: int | None = Field(default=None, ge=1)


class TrackUpdate(TrackCreate):
    """PUT 语义：整体替换一条进度。"""


class TrackOut(TrackCreate):
    model_config = ConfigDict(from_attributes=True)

    id: int
    source: str
    created_at: datetime
    updated_at: datetime


class TrackListOut(BaseModel):
    """列表 + 漏斗统计。统计随列表一起下发，免得前端为了画漏斗再拉一次全量。"""

    items: list[TrackOut]
    total: int
    # 各状态计数（含计数为 0 的状态），界面据此画漏斗。
    status_counts: dict[str, int] = Field(default_factory=dict)
    active_count: int
    offer_count: int
    rejected_count: int
    # 本月投递数：按 applied_at 落在当前自然月统计。
    month_count: int


class TrackMergePreview(BaseModel):
    """一条识别结果合并后会怎样——**不写库**，只回答"会发生什么"。"""

    record: TrackRecordIn
    action: str
    # 面向用户的原因，例如"已有记录处于「面试」，这条更早，保留现有进度"。
    reason: str
    # 合并前该条记录的当前状态；新建时为空串。
    current_status: str = ""


class TrackParseRequest(BaseModel):
    """粘贴的通知文本、通知截图或邮件导出文档（可同时给）。"""

    model_config = ConfigDict(extra="forbid")

    text: str = Field(default="", max_length=MAX_TRACK_TEXT_CHARS)
    images: list[ExtractionImageInput] = Field(
        default_factory=list, max_length=MAX_EXTRACTION_IMAGE_COUNT
    )
    documents: list[ExtractionDocumentInput] = Field(
        default_factory=list, max_length=MAX_EXTRACTION_DOCUMENT_COUNT
    )

    @model_validator(mode="after")
    def require_text_or_attachments(self) -> "TrackParseRequest":
        if not self.text.strip() and not self.images and not self.documents:
            raise ValueError("请粘贴通知内容，或上传至少一张截图或一份文档")
        return self


class TrackParseOut(BaseModel):
    items: list[TrackMergePreview]
    # ai / local：如实告诉用户这次是靠模型还是靠本地关键词匹配。
    parse_engine: str
    notes: list[str] = Field(default_factory=list)


class TrackApplyRequest(BaseModel):
    """用户确认后的写入。只提交勾选的条目。"""

    model_config = ConfigDict(extra="forbid")

    items: list[TrackRecordIn] = Field(default_factory=list, max_length=MAX_TRACK_RECORDS)
    # 这些条目的来源标记，默认「识别导入」；投递台自动记录时不走这个接口。
    source: str = Field(default="recognized", max_length=16)

    @field_validator("source")
    @classmethod
    def source_must_be_known(cls, value: str) -> str:
        from ..models.tracker import SOURCES

        cleaned = (value or "").strip() or "recognized"
        if cleaned not in SOURCES:
            raise ValueError(f"未知的记录来源「{cleaned}」")
        return cleaned


class TrackApplyItemResult(BaseModel):
    company: str
    title: str
    action: str
    status: str
    reason: str = ""


class TrackApplyOut(BaseModel):
    items: list[TrackApplyItemResult]
    created: int
    updated: int
    unchanged: int


__all__ = [
    "MAX_TRACK_RECORDS",
    "TrackApplyItemResult",
    "TrackApplyOut",
    "TrackApplyRequest",
    "TrackCreate",
    "TrackListOut",
    "TrackMergePreview",
    "TrackOut",
    "TrackParseOut",
    "TrackParseRequest",
    "TrackRecordIn",
    "TrackUpdate",
]

File: frontend/src/types/tracker.ts
/**
 * 求职进度类型。
 *
 * 取值与后端 `models/tracker.py` 的常量**逐字对应**——两处各写一份最容易漂移，
 * 而这里的漂移会表现成"界面上能选、保存时 422"。
 */

/** 漏斗的七个状态。顺序就是漏斗的推进顺序，「已结束」和「待确认」不在主线上。 */
export const TRACK_STATUSES = [
  "applied",
  "screening",
  "assessment",
  "interview",
  "offer",
  "rejected",
  "unknown",
] as const;
export type TrackStatus = (typeof TRACK_STATUSES)[number];

export const TRACK_STATUS_LABELS: Record<TrackStatus, string> = {
  applied: "已投递",
  screening: "筛选中",
  assessment: "测评/笔试",
  interview: "面试",
  offer: "Offer",
  rejected: "已结束",
  unknown: "待确认",
};

/** 漏斗主线（按推进顺序），「已结束」「待确认」不属于主线。 */
export const FUNNEL_STATUSES: readonly TrackStatus[] = [
  "applied",
  "screening",
  "assessment",
  "interview",
  "offer",
];

export const TRACK_STATUS_COLORS: Record<TrackStatus, string> = {
  applied: "default",
  screening: "blue",
  assessment: "cyan",
  interview: "purple",
  offer: "green",
  rejected: "red",
  unknown: "gold",
};

/** 进行中：还没走到终态的，界面上要突出。 */
export const ACTIVE_TRACK_STATUSES: readonly TrackStatus[] = [
  "applied",
  "screening",
  "assessment",
  "interview",
];

export function isActiveStatus(status: TrackStatus): boolean {
  return ACTIVE_TRACK_STATUSES.includes(status);
}

/** 记录来源：区分"投递台自动记的"和"我自己加的/识别导入的"。 */
export const TRACK_SOURCES = ["apply", "manual", "recognized"] as const;
export type TrackSource = (typeof TRACK_SOURCES)[number];

export const TRACK_SOURCE_LABELS: Record<TrackSource, string> = {
  apply: "投递台自动记录",
  manual: "手动添加",
  recognized: "识别导入",
};

/** 合并结论：这次识别对这条记录做了什么。 */
export const MERGE_ACTIONS = ["created", "updated", "unchanged"] as const;
export type MergeAction = (typeof MERGE_ACTIONS)[number];

export const MERGE_ACTION_LABELS: Record<MergeAction, string> = {
  created: "新增记录",
  updated: "更新进度",
  unchanged: "已是最新，未改动",
};

export const MERGE_ACTION_COLORS: Record<MergeAction, string> = {
  created: "green",
  updated: "blue",
  unchanged: "default",
};

export interface TrackRecord {
  company: string;
  title: string;
  status: TrackStatus;
  stage_note: string;
  applied_at: string;
  status_date: string;
  next_action: string;
  next_action_date: string;
  note: string;
  evidence: string;
}

export interface Track extends TrackRecord {
  id: number;
  source: TrackSource;
  created_at: string;
  updated_at: string;
  job_id: number | null;
  resume_id: number | null;
}

export type TrackPayload = Omit<Track, "id" | "source" | "created_at" | "updated_at">;

export interface TrackList {
  items: Track[];
  total: number;
  status_counts: Record<string, number>;
  active_count: number;
  offer_count: number;
  rejected_count: number;
  month_count: number;
}

/** 一条识别结果的合并预览：会新增、会更新，还是没变化。 */
export interface TrackMergePreview {
  record: TrackRecord;
  action: MergeAction;
  reason: string;
  current_status: string;
}

export interface TrackParseResult {
  items: TrackMergePreview[];
  parse_engine: "ai" | "local";
  notes: string[];
}

export interface TrackApplyResultItem {
  company: string;
  title: string;
  action: string;
  status: string;
  reason: string;
}

export interface TrackApplyResult {
  items: TrackApplyResultItem[];
  created: number;
  updated: number;
  unchanged: number;
}

export function emptyTrack(): TrackPayload {
  return {
    company: "",
    title: "",
    status: "applied",
    stage_note: "",
    applied_at: "",
    status_date: "",
    next_action: "",
    next_action_date: "",
    note: "",
    evidence: "",
    job_id: null,
    resume_id: null,
  };
}

/**
 * 从一条完整记录里取出可提交的字段（去掉服务端生成的 id、时间戳与来源）。
 *
 * 显式列字段而不是"解构后丢掉几个"：后者漏掉新字段时不会报错，只会静默把用户的
 * 输入丢掉——那是最难发现的一类毛病。
 */
export function trackPayload(track: Track, patch: Partial<TrackPayload> = {}): TrackPayload {
  return {
    company: track.company,
    title: track.title,
    status: track.status,
    stage_note: track.stage_note,
    applied_at: track.applied_at,
    status_date: track.status_date,
    next_action: track.next_action,
    next_action_date: track.next_action_date,
    note: track.note,
    evidence: track.evidence,
    job_id: track.job_id,
    resume_id: track.resume_id,
    ...patch,
  };
}
